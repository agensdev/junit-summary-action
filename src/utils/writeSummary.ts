import * as fs from "fs";
import convert from "xml-js";
import * as core from "@actions/core";
import { TestCases, TestCase } from "../classes/testCase";
import groupBy from "lodash.groupby";

const writeSummary = async (path: string, screenshots: Screenshot[]) => {
  const file = fs.readFileSync(path);

  const document = convert.xml2js(file.toString(), {
    compact: false,
  });

  const mergedTestCases = new TestCases();
  core.summary.addHeading("Test results");
  const testSuites: TestSuite[] = document.elements[0].elements;

  for (const element of testSuites) {
    const testSuitesElements = element.elements;

    testSuitesElements
      .filter((element) => element.name === "testcase")
      .forEach((testSuite) => {
        const errorMessage = testSuite.elements?.find(
          (element) => element.name === "failure"
        )?.attributes?.message;

        let status;

        if (!testSuite.elements) {
          status = "success";
        } else if (
          testSuite.elements.filter((element) => element.name === "skipped")
            .length === 1
        ) {
          status = "skipped";
        } else {
          status = "failed";
        }

        const test = new TestCase(
          element.attributes?.name,
          testSuite.attributes?.classname,
          testSuite.attributes?.name?.replace("()", ""),
          testSuite.attributes?.time,
          errorMessage,
          status
        );

        mergedTestCases.addTest(test);
      });
  }

  core.summary.addHeading("Summary", 2);
  core.summary.addTable([
    ["<strong>Overall result</strong>", mergedTestCases.testsEmoji],
    [
      "<strong>Total fails</strong>",
      mergedTestCases.numberOfFailedTests.toString(),
    ],
    [
      "<strong>Total skipped</strong>",
      mergedTestCases.numberOfSkippedTests.toString(),
    ],
    ["<strong>Total tests</strong>", mergedTestCases.numberOfTests.toString()],
  ]);

  // List skipped tests
  if (mergedTestCases.numberOfSkippedTests > 0) {
    core.summary.addHeading("Skipped tests", 2);
    mergedTestCases.skippedTests.forEach((test) => {
      core.summary.addTable([
        [
          `<h3 id='${test.id}'><a href="#${test.id}">ðŸ”—</a> <code>${test.fullTestName}</code></h3>`,
        ],
        [`<code>${test.errorMessage}</code>`],
      ]);
    });
  }

  // List failing tests
  if (mergedTestCases.numberOfFailedTests > 0) {
    core.summary.addHeading("Failing tests", 2);
    mergedTestCases.failedTests.forEach((test) => {
      const screenshot = screenshots
        ?.slice()
        .reverse()
        .find((screenshot) => {
          return (
            test.classname &&
            test.name &&
            screenshot.image.includes(test.classname) &&
            screenshot.image.includes(test.name + "()")
          );
        });
      core.summary.addTable([
        [
          `<h3 id='${test.id}'><a href="#${test.id}">ðŸ”—</a> <code>${test.fullTestName}</code></h3>`,
        ],
        [`<code>${test.errorMessage}</code>`],
      ]);
      screenshot && core.summary.addImage(screenshot.downloadUrl, "");
    });
  }

  // List all tests
  for (const testSuite of mergedTestCases.groupedTestsBySuite) {
    core.summary.addHeading(testSuite[0], 2);
    for (const [key, value] of groupedTests(testSuite[1])) {
      const testRows = value.map((test) => {
        return [
          {
            data: test.isSuccess
              ? test.name || ""
              : `<a href="#${test.id}">${test.name}</a>`,
            header: false,
          },
          { data: test.timeFormatted, header: false },
          { data: test.testEmoji, header: false },
        ];
      });
      core.summary.addHeading(key, 3);
      core.summary.addTable(headerRow.concat(testRows));
    }
  }

  core.setOutput("total", mergedTestCases.numberOfTests);
  core.setOutput("passed", mergedTestCases.numberOfPassedTests);
  core.setOutput("failed", mergedTestCases.numberOfFailedTests);
  core.setOutput("skipped", mergedTestCases.numberOfSkippedTests);
  core.setOutput("testSummary", mergedTestCases.testSummary);
  core.setOutput("passPercent", mergedTestCases.passPercent);

  await core.summary.write();
};

const headerRow = [
  [
    { data: "Test name", header: true },
    { data: "Time elapsed", header: true },
    { data: "Result", header: true },
  ],
];

const groupedTests = (tests: TestCase[]) =>
  Object.entries(groupBy(tests, "classname"));

export default writeSummary;
