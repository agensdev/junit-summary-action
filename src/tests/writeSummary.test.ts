import writeSummary from "../utils/writeSummary";
import * as core from "@actions/core";
import {
  expect,
  jest,
  describe,
  beforeEach,
  afterEach,
  it,
} from "@jest/globals";

jest
  .spyOn(core.summary, "write")
  .mockReturnValue(Promise.resolve("test" as any));

const mockXmlFail = `<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="1" failures="1">
    <testsuite name="ExampleTests" tests="1" failures="1" skipped="0" time="65.83285999298096">
        <properties>
          <property name="Configuration" value="Configuration 1"/>
        </properties>
        <testcase classname="ExampleUITests" name="testExample()" time="10.829154968261719">
            <failure message="XCTAssertEqual failed: (&quot;expected result&quot;) is not equal to (&quot;unexpected result&quot;) (/path/to/test.swift#CharacterRangeLen=0&amp;EndingLineNumber=15&amp;StartingLineNumber=15)">
            </failure>
        </testcase>
    </testsuite>
</testsuites>`;

const mockXmlSuccess = `<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="jest tests" tests="4" failures="0" errors="0" time="5.004">
  <testsuite name="uploadScreenshots" errors="0" failures="0" skipped="0" timestamp="2023-11-06T09:52:41" time="4.435" tests="2">
    <testcase classname="uploadScreenshots throws an error if the Firebase service account is missing" name="uploadScreenshots throws an error if the Firebase service account is missing" time="0.015">
    </testcase>
    <testcase classname="uploadScreenshots successfully uploads screenshots and returns download URLs" name="uploadScreenshots successfully uploads screenshots and returns download URLs" time="0.011">
    </testcase>
  </testsuite>
  <testsuite name="writeSummary" errors="0" failures="0" skipped="0" timestamp="2023-11-06T09:52:46" time="0.476" tests="2">
    <testcase classname="writeSummary should write the summary correctly, no screenshots" name="writeSummary should write the summary correctly, no screenshots" time="0.012">
    </testcase>
    <testcase classname="writeSummary should write the summary correctly, with screenshots" name="writeSummary should write the summary correctly, with screenshots" time="0.003">
    </testcase>
  </testsuite>
</testsuites>
`;

jest.mock("fs", () => ({
  readFileSync: jest.fn().mockImplementation((path) => {
    if ((path as string).includes("fail")) {
      return mockXmlFail;
    } else {
      return mockXmlSuccess;
    }
  }),
  existsSync: jest.fn(),
  promises: {
    access: jest.fn().mockImplementation(() => Promise.resolve()),
  },
}));

describe("writeSummary", () => {
  beforeEach(async () => {
    jest.clearAllMocks();
    await core.summary.clear();
    process.env.GITHUB_STEP_SUMMARY = "./debug/summary.md";
  });

  afterEach(() => {
    delete process.env.GITHUB_STEP_SUMMARY;
    jest.clearAllMocks();
  });

  it("should write the summary correctly, failure, no screenshots", async () => {
    await writeSummary("fail.xml", []);
    expect(core.summary).toMatchSnapshot();
  });

  it("should write the summary correctly, failure, with screenshots", async () => {
    const screenshots: Screenshot[] = [
      {
        image: "ExampleUITests.testExample()",
        downloadUrl: "https://url.to/img",
      },
    ];

    await writeSummary("./fail.xml", screenshots);
    expect(core.summary).toMatchSnapshot();
  });

  it("should write the summary correctly, success, no screenshots", async () => {
    await writeSummary("success.xml", []);
    expect(core.summary).toMatchSnapshot();
  });
});
