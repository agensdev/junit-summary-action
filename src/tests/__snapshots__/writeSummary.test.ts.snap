// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeSummary should write the summary correctly, failing variation 1, no screenshots 1`] = `
Summary {
  "_buffer": "<h1>Test results</h1>
<h2>Summary</h2>
<table><tr><td><strong>Overall result</strong></td><td>‚ùå</td></tr><tr><td><strong>Total fails</strong></td><td>1</td></tr><tr><td><strong>Total skipped</strong></td><td>0</td></tr><tr><td><strong>Total tests</strong></td><td>4</td></tr></table>
<h2>Failing tests</h2>
<table><tr><td><h3 id='lottery results-lottery results'><a href="#lottery results-lottery results">üîó</a> <code>Lottery results - Lottery results</code></h3></td></tr><tr><td><code>Assertion is false: "Hver tirsdag og fredag!" is visible</code></td></tr></table>
<h2>Hjem</h2>
<h3>Profil Name</h3>
<table><tr><th>Test name</th><th>Time elapsed</th><th>Result</th></tr><tr><td>Profil Name</td><td>11.0 s</td><td>‚úÖ</td></tr></table>
<h3>Nabolaget - Address is shown</h3>
<table><tr><th>Test name</th><th>Time elapsed</th><th>Result</th></tr><tr><td>Nabolaget - Address is shown</td><td>14.0 s</td><td>‚úÖ</td></tr></table>
<h3>Lottery results</h3>
<table><tr><th>Test name</th><th>Time elapsed</th><th>Result</th></tr><tr><td><a href="#lottery results-lottery results">Lottery results</a></td><td>69.0 s</td><td>‚ùå</td></tr></table>
<h3>Eurojackpot - Andelsbank feature toggle</h3>
<table><tr><th>Test name</th><th>Time elapsed</th><th>Result</th></tr><tr><td>Eurojackpot - Andelsbank feature toggle</td><td>18.0 s</td><td>‚úÖ</td></tr></table>
",
  "write": [MockFunction] {
    "calls": [
      [
        {
          "overwrite": true,
        },
      ],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
}
`;

exports[`writeSummary should write the summary correctly, failure, no screenshots 1`] = `
Summary {
  "_buffer": "<h1>Test results</h1>
<h2>Summary</h2>
<table><tr><td><strong>Overall result</strong></td><td>‚ùå</td></tr><tr><td><strong>Total fails</strong></td><td>1</td></tr><tr><td><strong>Total skipped</strong></td><td>0</td></tr><tr><td><strong>Total tests</strong></td><td>1</td></tr></table>
<h2>Failing tests</h2>
<table><tr><td><h3 id='exampleuitests-testexample'><a href="#exampleuitests-testexample">üîó</a> <code>ExampleUITests - testExample</code></h3></td></tr><tr><td><code>XCTAssertEqual failed: ("expected result") is not equal to ("unexpected result") (/path/to/test.swift#CharacterRangeLen=0&EndingLineNumber=15&StartingLineNumber=15)</code></td></tr></table>
<h2>ExampleTests</h2>
<h3>ExampleUITests</h3>
<table><tr><th>Test name</th><th>Time elapsed</th><th>Result</th></tr><tr><td><a href="#exampleuitests-testexample">testExample</a></td><td>10.8 s</td><td>‚ùå</td></tr></table>
",
  "write": [MockFunction] {
    "calls": [
      [
        {
          "overwrite": true,
        },
      ],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
}
`;

exports[`writeSummary should write the summary correctly, failure, with screenshots 1`] = `
Summary {
  "_buffer": "<h1>Test results</h1>
<h2>Summary</h2>
<table><tr><td><strong>Overall result</strong></td><td>‚ùå</td></tr><tr><td><strong>Total fails</strong></td><td>1</td></tr><tr><td><strong>Total skipped</strong></td><td>0</td></tr><tr><td><strong>Total tests</strong></td><td>1</td></tr></table>
<h2>Failing tests</h2>
<table><tr><td><h3 id='exampleuitests-testexample'><a href="#exampleuitests-testexample">üîó</a> <code>ExampleUITests - testExample</code></h3></td></tr><tr><td><code>XCTAssertEqual failed: ("expected result") is not equal to ("unexpected result") (/path/to/test.swift#CharacterRangeLen=0&EndingLineNumber=15&StartingLineNumber=15)</code></td></tr></table>
<img src="https://url.to/img" alt="">
<h2>ExampleTests</h2>
<h3>ExampleUITests</h3>
<table><tr><th>Test name</th><th>Time elapsed</th><th>Result</th></tr><tr><td><a href="#exampleuitests-testexample">testExample</a></td><td>10.8 s</td><td>‚ùå</td></tr></table>
",
  "write": [MockFunction] {
    "calls": [
      [
        {
          "overwrite": true,
        },
      ],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
}
`;

exports[`writeSummary should write the summary correctly, success, no screenshots 1`] = `
Summary {
  "_buffer": "<h1>Test results</h1>
<h2>Summary</h2>
<table><tr><td><strong>Overall result</strong></td><td>‚úÖ</td></tr><tr><td><strong>Total fails</strong></td><td>0</td></tr><tr><td><strong>Total skipped</strong></td><td>0</td></tr><tr><td><strong>Total tests</strong></td><td>4</td></tr></table>
<h2>uploadScreenshots</h2>
<h3>uploadScreenshots throws an error if the Firebase service account is missing</h3>
<table><tr><th>Test name</th><th>Time elapsed</th><th>Result</th></tr><tr><td>uploadScreenshots throws an error if the Firebase service account is missing</td><td>0.0 s</td><td>‚úÖ</td></tr></table>
<h3>uploadScreenshots successfully uploads screenshots and returns download URLs</h3>
<table><tr><th>Test name</th><th>Time elapsed</th><th>Result</th></tr><tr><td>uploadScreenshots successfully uploads screenshots and returns download URLs</td><td>0.0 s</td><td>‚úÖ</td></tr></table>
<h2>writeSummary</h2>
<h3>writeSummary should write the summary correctly, no screenshots</h3>
<table><tr><th>Test name</th><th>Time elapsed</th><th>Result</th></tr><tr><td>writeSummary should write the summary correctly, no screenshots</td><td>0.0 s</td><td>‚úÖ</td></tr></table>
<h3>writeSummary should write the summary correctly, with screenshots</h3>
<table><tr><th>Test name</th><th>Time elapsed</th><th>Result</th></tr><tr><td>writeSummary should write the summary correctly, with screenshots</td><td>0.0 s</td><td>‚úÖ</td></tr></table>
",
  "write": [MockFunction] {
    "calls": [
      [
        {
          "overwrite": true,
        },
      ],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
}
`;
